### ISSUE TITLE
Rule Change: camelcase allow snake when chaining

### ISSUE THREAD

**Issue by pajakpawel** [eslint/eslint#16604]
### What rule do you want to change?

camelcase

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
this.data.nested.variable_from_backend = 'any value'
```

### What does the rule currently do for this code?

Currently, rule highlights `variable_from_backend` name as violation of camelcase rule.
The problem is only the `variable_from_backend` is a violation and not the whole `this.data.nested.variable_from_backend`, because if thole chaining would be treated as violation it would be possible to use `allow` option and regex `(?<=\.)[a-z_]*` to allow all snake_case names that appear after a dot.

### What will the rule do after it's changed?

I propose the rule provide some option to allow snake_case if it is in chaining notation.
This is due to the fact, that some variables in external services might require snake_case.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

**Comment by mdjermanovic** [comment]
Hi @pajakpawel, thanks for the issue!

> I propose the rule provide some option to allow snake_case if it is in chaining notation.
> This is due to the fact, that some variables in external services might require snake_case.

This is already covered by the `properties: "never"` option. Here's a [playground demo](https://eslint.org/play/#eyJ0ZXh0IjoiLyogZXNsaW50IGNhbWVsY2FzZTogW1wiZXJyb3JcIiwgeyBcInByb3BlcnRpZXNcIjogXCJuZXZlclwiIH1dICovXG5cbnRoaXMuZGF0YS5uZXN0ZWQudmFyaWFibGVfZnJvbV9iYWNrZW5kID0gJ2FueSB2YWx1ZScgLy8gbm8gZXJyb3IiLCJvcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYVZlcnNpb24iOiJsYXRlc3QiLCJzb3VyY2VUeXBlIjoic2NyaXB0IiwiZWNtYUZlYXR1cmVzIjp7fX0sInJ1bGVzIjp7fSwiZW52Ijp7ImVzNiI6dHJ1ZX19fQ==).

**Comment by pajakpawel** [comment]
Thanks for your response @mdjermanovic!
Unfortunately, I can't agree, that `properties: "never"` covers my issue, because it also allows for newly created properties to be in notation other than `camelCase`. Consider example:
```javascript
const someObject = { new_property: 'value' }
```
Using `properties: "never"` will allow for that `new_property` written in snake_case. The desired behavior would be to raise error when new property is created in other than `camelCase` notation, while allowing for `snake_case` in chaining.

**Comment by mdjermanovic** [comment]
Property assignments can also create new properties (for example, `const someObject = {}; someObject.new_property = 'value';`) and are therefore included in the same option. I think there's no difference between property assignments and object literals in regard to the purpose of this option, which is to allow non-camelcase property names as they might be required by external APIs. If you would like more granular options, you can copy the core rule into a custom rule and modify it.

**Comment by pajakpawel** [comment]
Ok, thanks for the clarification and the idea of copying core rule to custom rule and modifying!

I've also found that satisfying an external API, while keeping `camelcase` rule ON, might be accomplished by using `allowPattern` option of  `dot-notation` rule, maybe it would be a good idea to add such use case in `camelcase` rule documentation.

I mean something like "If you wish to use other notation just to satisfy external API, look here <link to `dot-notation` rule's `alowPattern` option".
It is tricky to look for solution in very different rule.
